### ISSUE TITLE
GeoJSON comments

### ISSUE THREAD

**Issue by voidqk** [voidqk/polybooljs#7]
Please leave comments here about the experimental GeoJSON support.

**Comment by voidqk** [comment]
#11 says GeoJSON is working as expected.

**Comment by Portur** [comment]
The documentation is a bit unclear for GeoJson. 

From the function call i'm assuming polygonFromGeoJSON will retrieve a single polygon from a geojson object that only contains a single polygon?

Also the inverted tag has no explanation so I am unsure if this is auto generated when reading from geojson? Maybe an example for this would help.

My use case is to combine 10 to 500 polygons retrieved from 4-15 geojson files. Do I need to extract them individually to union them? polygonFromGeoJSON seems to be what I'm looking for but I'm not sure if the result will be what I am expecting it to do, which is extract all polygon(s) and return an array of PolyboolJS polygons to use in union/intersect/etc.

**Comment by voidqk** [comment]
The `polygonFromGeoJSON` function should work on `Polygon` and `MultiPolygon` GeoJSON types.  It should convert a GeoJSON object to a polybooljs polygon.

The easiest way to explain the inverted flag is to see it in the demo:

https://rawgit.com/voidqk/polybooljs/master/dist/demo.html

If you click on the Invert Red or Invert Blue buttons, you can see what toggling the invert flag does.  What's considered outside the polygon is now considered inside, and vice-versa.

I don't know if you need to extract the GeoJSON individually, I don't know what your data looks like.  The `polygonFromGeoJSON` function takes an object that looks like `{ type: "MultiPolygon", coordinates: [[...]] }`.

**Comment by Portur** [comment]
Extract all polygons from a geojson file. eg.

`{
    "type": "Feature",
    "properties": {"party": "Republican"},
    "geometry": {
        "type": "Polygon",
        "coordinates": [[
            [-104.05, 48.99],
            [-97.22,  48.98],
            [-96.58,  45.94],
            [-104.03, 45.94],
            [-104.05, 48.99]
        ]]
    }
}, {
    "type": "Feature",
    "properties": {"party": "Democrat"},
    "geometry": {
        "type": "Polygon",
        "coordinates": [[
            [-109.05, 41.00],
            [-102.06, 40.99],
            [-102.03, 36.99],
            [-109.04, 36.99],
            [-109.05, 41.00]
        ]]
    }
},{...}`

which can be up to `n` polygons.

Will `polygonFromGeoJSON` only extract the firslt polygon?

About the inverted tag; Will the resulting poly from `polygonFromGeoJSON` contain a `default` eg `{inverted : false}` ? Assuming most polys are not inverted.

**Comment by Portur** [comment]
A handy page to play with geojson

[http://bl.ocks.org/d/25f1d4681f44f825f06000292109dc4a](http://bl.ocks.org/d/25f1d4681f44f825f06000292109dc4a)

**Comment by voidqk** [comment]
Right, so the object that has `"type": "Polygon"` is what needs to be passed to `polygonFromGeoJSON`, i.e., `data[n].geometry`.

Yes, `polygonFromGeoJSON` will fill out the inverted flag.

**Comment by cyrilchapon** [comment]
Hey.

Just to add some random feedbacks and caveats of my knowledge and experience about GeoJSON, maybe could help...

- The `types` field should respect case (and is camel cased)
- A **Feature** is something in the shape `{type: 'Feature', properties: {}, geometry: <Geometry>}`
  - `type`, `properties` (don't forget it) and `geometry` are mandatory
- A **Geometry** is something in the shape `{type: <GeometryType>, coordinates: []}` **OR** `{type: 'GeometryCollection', geometries: [Geometry]}`
  - There is a trick here, and this library should be concerned about this : `GeometryCollection` **is** a valid geometry, and thus changes the shape of the root Geometry object...
  - I don't know if a GeometryCollection can be a geometry child of another GeometryCollection, but that's worth checking it
- About coordinates, they are separated in "rings" logic
  - A Polygon `coordinates` member is in the shape
  ```
  [ // <= Array of 'rings'
    [ // Single outer 'ring'. It MUST be counter-clockwise
      [lng, lat], [lng, lat], [...] // Coordinates serie MUST have last === first
    ],
    [ // Inner 'ring' (hole). It MUST be clockwise
      [lng, lat], [lng, lat], [...] // Coordinates serie MUST have last === first
    ],
    [ // Another inner 'ring' (hole). It MUST be clockwise
      [lng, lat], [lng, lat], [...] // Coordinates serie MUST have last === first
    ]
  ]
  ```
  - I don't know if a "hole" (inner ring) can contain an "island". But that's worth checking...
  - A Position (coordinates couple) is **not limited** to a length of 2, and could include, eg., elevation as a 3rd member (`[lng, lat, elevation]`)


Some useful stuff :

- [geojson-toolkit](https://github.com/dmtucker/gjtk-js)
- [turfjs](http://turfjs.org/)

**Comment by Sagarpreet** [comment]
So does this library returns holes when taking union ?

**Comment by domiSchenk** [comment]
Just switched from Turf to PolyBoolJs, and i must say i love it! We had many problems with Turf when working with it. Sometimes the resulting geometry is invalid, sometimes it create unreasonable geometries.

With PolyBoolJs we didn't encounter such problems!

What would be nice is if we could pass a Feature to PolyBoolJs and not only a geometry. Also if you stick to geometry a more meaningful error would be nice. I first thought the GeoJSON addon did not work.

An Error like: 
`PolyBool: You didn't pass a GeoJSON geometry to PolyBool`
Or
`PolyBool does not support Type: {type}. Try passing a geometry of type 'Polygon' or 'MultiPoligon'`
would be nice

Keep up the good work!
